reader.go

```diff
@@ -484,28 +1462,47 @@ func (r *Reader) readLag(ctx context.Context) {
 	}
 }
 
-func (r *Reader) start() {
+func (r *Reader) start(offsetsByPartition map[int]int64) {
 	ctx, cancel := context.WithCancel(context.Background())
 
 	r.cancel() // always cancel the previous reader
 	r.cancel = cancel
 	r.version++
 
-	r.join.Add(1)
-	go (&reader{
-		dialer:      r.config.Dialer,
-		logger:      r.config.Logger,
-		errorLogger: r.config.ErrorLogger,
-		brokers:     r.config.Brokers,
-		topic:       r.config.Topic,
-		partition:   r.config.Partition,
-		minBytes:    r.config.MinBytes,
-		maxBytes:    r.config.MaxBytes,
-		maxWait:     r.config.MaxWait,
-		version:     r.version,
-		msgs:        r.msgs,
-		stats:       &r.stats,
-	}).run(ctx, r.offset, &r.join)
+	r.join.Add(len(offsetsByPartition))
+	for partition, offset := range offsetsByPartition {
+		go func(ctx context.Context, partition int, offset int64, join *sync.WaitGroup) {
+			defer join.Done()
+
+			(&reader{
+				dialer:      r.config.Dialer,
+				logger:      r.config.Logger,
+				errorLogger: r.config.ErrorLogger,
+				brokers:     r.config.Brokers,
+				topic:       r.config.Topic,
+				partition:   partition,
+				minBytes:    r.config.MinBytes,
+				maxBytes:    r.config.MaxBytes,
+				maxWait:     r.config.MaxWait,
+				version:     r.version,
+				msgs:        r.msgs,
+				stats:       &r.stats,
+			}).run(ctx, offset)
+		}(ctx, partition, offset, &r.join)
+	}
+}
+
+func (r *Reader) CommitMessage(m Message) error {
```

Comment:
Should we do this automatically as part of the `ReadMessage` method? I feel like otherwise a program has keep awareness of whether a Reader is part of a consumer group to know if it must call `CommitMessage` or not.

I feel like managing the commits automatically would better be inline with the current design where the offsets are already automatically managed when directly addressing partitions.

If this is a critical feature for one of your systems I think we should introduce a new method (say `PeekMessage`) which doesn't do automatic offset management and leaves it up to the program to explicitly commit messages.

Reply:
Perhaps we could go through some use cases.  

For us, it is critical that messages are processed before we commit the message off the Reader which requires an explicit commit.  Consider what would happen if one were performing db replication or event sourcing if a message that was read for reason failed to be processed.  Instead of INSERT > UPDATE > DELETE we could end up with just INSERT > UPDATE

That said, I can see some cases where keeping track of the offsets might not be as critical and a lossy, automatic commit would be acceptable.  Perhaps an AutoCommit flag on ReaderConfig?

Reply:
Do you have a concern about adding a new method that doesn’t do the auto-commit?

I feel like that would make it explicit whether CommitMessage has to be called or not. I would really prefer if we can keep the auto-commit behavior in ReadMessage. 

Also what do you think about changing the commit method to be called CommitMessages and accept an arbitrary number of messages as argument? That way the program acquired some ownership over the batching of commits (if we can make the commits synchronized I can see this being useful for some of our use cases at Segment, but we must get the API right to allow this).

Reply:
PS: what do you think about passing a context as well to the CommitMeasages method? That way we also give control to the application on how long it may wait for the commit to be published?

Reply:
I think I'm just very used to thinking of read messages as being a singular activity and autocommit as a property of the connection rather than having ReadMessage vs PeekMessage/CommitMessage.  How would you feel about FetchMessage/CommitMessage for naming which would be more in sync with Kafka's naming.

Reply:
Here are the commit use cases that I'm thinking of.

1. fetch message, process message, commit message used by most of our use cases.  Since we require all services support at least once processing, async commits are fine since replays will be ignored.

2. fetch/process/commit for exactly once processing.  I'm not sure this is a thing with the api as we have it.  I believe Confluent created a new api to support these semantics, but admittedly I haven't read up on it.  Exactly once processing would really have to be at least once since there's no two phase commit.

3. fetch/process/batch-commit for micro-batch processing.

4. fetch/process/no-commit e.g. live statistics.  

4 doesn't require a commit, so that's easy.  1 & 2 only require a single message commit where 3 requires an arbitrarily sized batch.  

What about having two commit methods?  Sync/async could still be controlled by the CommitInterval field.

```
Commit(ctx context.Context, offsets Offsets) error
CommitMessage(m kafka.Message)
```

1 & 2 seems like they would almost always want to an async commit with a single message.  Having a ctx seems like it would be confusing.  

3 seems like it could benefit from a sync commit commit.  By separating the calls, we could also separate the shape of the data being sent.

For comparison, sarama-cluster uses the following:

```
func (c *Consumer) MarkOffset(msg *sarama.ConsumerMessage, metadata string)
func (c *Consumer) MarkOffsets(s *OffsetStash)
```

Reply:
What I really want to avoid is creating a dependency between how the Reader is configured and how it is used, which is why I would prefer to have auto commits on ReadMessage. That way, no matter what the configuration is, the code does not need to be changed. Otherwise, the code using the Reader needs to be aware of its configuration to know whether it has to commit or not, or it has to maybe ignore the error, or we make the commit a no-op if it’s not needed? All those patterns seem less robust than splitting the APIs to have the program explicitly control how it consumes messages when it needs to.

FetchMessage sounds good to me (I actually like it better than Peek).

I feel like making the commit sync or async can be configured on the Reader, because in this case it doesn’t impact the way the program uses the code, it only makes a blocking operation non-blocking. This is actually the approach we took in the Writer to configure whether the writes are done synchronously it asynchronously. 

I’m not concerned about passing a context for cases where commits are asynchronous, very likely the program would use the same context that it passed to FetchMessage.

I’d like to avoid introducing a new Offset concept here, I liked your proposal of passing the message to commit, it makes it trivial to use the API, otherwise the program needs to extract the offsets etc...

It seems to me that a method with this signature would cover all use cases:
```go
CommitMessages(ctx context.Context, msg ...Message) error
```
(Plus it’s consistent with the model used with WriteMessages in the Writer)

Reply:
Works for me.